cmd/snap-recovery-chooser/main_test.go

```diff
@@ -0,0 +1,466 @@
+	c.Assert(rsp, IsNil)
+}
+
+func (s *cmdSuite) TestRunUIToolErr(c *C) {
+	mockCmd := testutil.MockCommand(c, "tool", `
+echo foo
+exit 22
+`)
+	defer mockCmd.Restore()
+
+	_, err := main.RunUI(exec.Command(mockCmd.Exe()), mockSystems)
+	c.Assert(err, ErrorMatches, "cannot collect output of the UI process: exit status 22")
+}
+
+func (s *cmdSuite) TestRunUIInputJSON(c *C) {
+	d := c.MkDir()
+	tf := filepath.Join(d, "json-input")
+	mockCmd := testutil.MockCommand(c, "tool", fmt.Sprintf(`
+cat > %s
+echo '{}'
+`, tf))
+	defer mockCmd.Restore()
+
+	_, err := main.RunUI(exec.Command(mockCmd.Exe()), mockSystems)
+	c.Assert(err, IsNil)
+
+	data, err := ioutil.ReadFile(tf)
+	c.Assert(err, IsNil)
+	var input *main.ChooserSystems
+	err = json.Unmarshal(data, &input)
+	c.Assert(err, IsNil)
+
+	c.Assert(input, DeepEquals, mockSystems)
+}
+
+func (s *cmdSuite) TestStdoutUI(c *C) {
+	var buf bytes.Buffer
+	err := main.OutputForUI(&buf, mockSystems)
+	c.Assert(err, IsNil)
+
+	var out *main.ChooserSystems
+
+	err = json.Unmarshal(buf.Bytes(), &out)
+	c.Assert(err, IsNil)
+	c.Assert(out, DeepEquals, mockSystems)
+}
+
+type mockedClientCmdSuite struct {
+	baseCmdSuite
+
+	config client.Config
+}
+
+var _ = Suite(&mockedClientCmdSuite{})
+
+func (s *mockedClientCmdSuite) SetUpTest(c *C) {
+	s.baseCmdSuite.SetUpTest(c)
+}
+
+func (s *mockedClientCmdSuite) RedirectClientToTestServer(handler func(http.ResponseWriter, *http.Request)) {
+	server := httptest.NewServer(http.HandlerFunc(handler))
+	s.BaseTest.AddCleanup(func() { server.Close() })
+	s.config.BaseURL = server.URL
+}
+
+type mockSystemRequestResponse struct {
+	label  string
+	code   int
+	expect map[string]interface{}
+}
+
+func (s *mockedClientCmdSuite) mockSuccessfulResponse(c *C, rspSystems *main.ChooserSystems, reqSystemLabel *mockSystemRequestResponse) {
+	n := 0
+	s.RedirectClientToTestServer(func(w http.ResponseWriter, r *http.Request) {
+		switch n {
+		case 0:
+			c.Check(r.URL.Path, Equals, "/v2/systems")
+			err := json.NewEncoder(w).Encode(apiResponse{
+				Type:       "sync",
+				Result:     rspSystems,
+				StatusCode: 200,
+			})
+			c.Assert(err, IsNil)
+		case 1:
+			if reqSystemLabel == nil {
+				c.Fatalf("unexpected request to %q", r.URL.Path)
+			}
+			c.Check(r.URL.Path, Equals, "/v2/systems/"+reqSystemLabel.label)
+			c.Check(r.Method, Equals, "POST")
+
+			var data map[string]interface{}
+			err := json.NewDecoder(r.Body).Decode(&data)
+			c.Assert(err, IsNil)
+			c.Check(data, DeepEquals, reqSystemLabel.expect)
+
+			rspType := "sync"
+			var rspData map[string]string
+			if reqSystemLabel.code >= 400 {
+				rspType = "error"
+				rspData = map[string]string{"message": "failed in mock"}
+			}
+			err = json.NewEncoder(w).Encode(apiResponse{
+				Type:       rspType,
+				Result:     rspData,
+				StatusCode: reqSystemLabel.code,
+			})
+			c.Assert(err, IsNil)
+		default:
+			c.Fatalf("expected to get 1 requests, now on %d", n+1)
```

Comment:
Nitpick, the number here is zero-based, which might be surprising if the test ever fails and someone has to debug it.

Reply:
It's a pattern we use throughout the client/snap command tests.